V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-RESILIENCE/MASWE-0052.md

Reviewer: DONE

---
title: Resource Obfuscation Not Implemented
id: MASWE-0052
alias: resource-obfuscation
requirement: "The app applies resource obfuscation to hinder reverse engineering."
platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0052
attacks: [MAS-ATTACK-0001]
mappings:
  masvs-v1: [MSTG-RESILIENCE-11]
  masvs-v2: [MASVS-RESILIENCE-3]
  cwe: [693]
  maswe-beta: [MASWE-0090]
status: new
---

## Overview

This weakness occurs when an app's resources and assets are left in clear, unprotected form, revealing how the app works and aiding reverse engineering.

> Beyond code, strings, layouts, configuration files, bundled data, and native binaries all carry information about the app's inner workings. Attackers mining the package can use them to map features, find security-relevant configuration, and identify targets for tampering. Note that obfuscation or encryption applied without integrity validation can itself be tampered with, so resource protection should complement, not replace, integrity checks (see @MASWE-0063).

Reviewer: ```suggestion
In addition to application code, various artifacts such as strings, layouts, configuration files and bundled data all contain information about the app's inner workings. Attackers mining the package can use them to map features, find security-relevant configuration, and identify targets for tampering. Note that obfuscation or encryption applied without integrity validation can itself be tampered with, so resource protection should complement, not replace, integrity checks (see @MASWE-0063).
```

Reviewer: Note that obfuscation or encryption applied without integrity validation can itself be tampered with, so resource protection should complement, not replace, integrity checks (see @MASWE-0063). -> If we put thiss in 0052 (this one, resource obfuscation), we should also put this in 0051 (code obfuscation).

## Modes of Introduction

> - **Resources Left in Clear**: Shipping strings, configuration, and other assets unobfuscated and unencrypted in the package.

Reviewer: ```suggestion
- **Resources Left in Clear**: Shipping strings, configuration, and other assets unobfuscated and unencrypted in the application package.
```

> - **Native Binaries Not Protected**: Shipping native libraries without encryption or packing, exposing their full structure to static analysis.

Reviewer: ```suggestion
```

Reviewer: This goes in 0051, code obfuscation.

- **Identifiers Left Meaningful**: Keeping descriptive resource and string identifiers that map directly to features and security controls.

> - **Protection Without Integrity Validation**: Obfuscating or encrypting resources without verifying their integrity, leaving the protected resources silently replaceable.

Reviewer: ```suggestion
```

Reviewer: This MASWE should focus on obfuscation, other MASWE should focus on integrity validation. I think this mixes both and might not be the best option.

## Impact

> - **Bypass of Protection Mechanisms**: Attackers can locate security-relevant configuration and assets and use them to target the app's defenses, resulting in easier circumvention of its protections.

Reviewer: ```suggestion
- **Bypass of Protection Mechanisms**: Attackers can locate security-relevant configuration and assets and use cross-references to target the app's security-relevant code locations, resulting in easier circumvention of its protections.
```

- **Compromise of Sensitive Data**: Attackers can extract proprietary content and internal information from clear resources, resulting in intellectual property exposure.

## Mitigations

> - **Obfuscate or Encrypt Sensitive Resources**: Protect strings, configuration, and assets that reveal security-relevant behavior, decrypting them only when needed at runtime.

Reviewer: ```suggestion
- **Obfuscate Sensitive Resources**: Obfuscate strings, configuration, and assets that reveal security-relevant behavior.
```

Reviewer: Resource encryption is obfuscation

> - **Pack or Encrypt Native Binaries**: Apply packing or encryption to native libraries containing sensitive logic.

Reviewer: ```suggestion
```

Reviewer: Again, this is in code obfuscation (0051)
